### PR TITLE
capabilities API, no hook-in to route management

### LINKED ISSUE #22: User capabilities
Build user capability system.

Per conference, user has any of `['read', 'write', 'admin']` which roughly correspond to attendees, organizers, and admins.

### PR DESCRIPTION
**Pull request by TheKevJames**
Fixes #22.

### CODE REVIEW

**Review by kyl191 — COMMENTED** [review]
Waiting for @schneider-clarisse & @larajanecka to explain the use-case, it looks like we can just pull it in directly from auth0 instead of storing it ourselves.

**Comment by larajanecka** [comment]
Theres no reason we can't. The main concern is that the backend also needs to know the user's permission so however it gets that data has to be secure. I'm guessing that passing it from the front end in a request is the wrong way to do it.

File: database/structure.sql [hunk 1]
```diff
@@ -51,3 +51,35 @@ CREATE TABLE broadcast (
     message         text        NOT NULL,
     expiryTime      bigint      NOT NULL
 );
+
+DROP TABLE IF EXISTS capability CASCADE;
+CREATE TABLE capability (
+    id              text        PRIMARY KEY REFERENCES account(id) ON DELETE CASCADE,
+    read            boolean     NOT NULL DEFAULT TRUE,
```
Comment (kyl191) [inline database/structure.sql]: @TheKevJames Can we use an enum (read/write/admin) and only store the user id of the special users? If there's no matching ID found, default to read-only?

File: database/structure.sql [hunk 2]
```diff
@@ -51,3 +51,35 @@ CREATE TABLE broadcast (
...
+$$
+BEGIN
+    DELETE from capability WHERE id = OLD.id;
+    RETURN NEW;
+END;
+$$
+LANGUAGE 'plpgsql';
+
+CREATE TRIGGER account_capability_delete_trigger AFTER DELETE ON account
+    FOR EACH ROW EXECUTE PROCEDURE account_capability_delete();
```
Comment (kyl191) [inline database/structure.sql]: Why not just add a column to the account table instead?

**Comment by kyl191** [comment]
Well in this case we're already trusting the token from Auth0, so trusting an extra field from the token isn't much of a worry. For the front-end, could we just query Auth0 directly?

We're slowly building a shadow user db, storing everything but credentials. I've been treating it as essentially a cache which can be thrown away and repopulated from Auth0 for things like mapping userid <-> actual names. The Twitter & FB cross-posting fetches the user's credentials from Auth0 at runtime and throws them away once the post is made.

We'll now have two sources of truth - admittedly for different properties, but we'll inevitably start adding stuff onto our database. Are we OK with this?

**Comment by TheKevJames** [comment]
@kyl191 Are you suggesting the API trust whatever access level the frontend tells us it is on a request. That seems.. brittle. 

```
curl -XDELETE www.calligre.com/api/user/1?my-access-level=root
```

**Comment by kyl191** [comment]
No, I suggesting the API trust whatever capability is listed & signed in the JWT. Client editable is a definite no, but the JWT is generated by a trusted third party, and signing it prevents clients from editing it.

**Comment by TheKevJames** [comment]
auth0 doesn't support adding additional fields to the JWT itself, it supports user properties stored on their system and accessible by API call. ie:

```
token = un_jwt_ify(frontend_data)
response = requests.get('auth0.com/route/to/user_info', token)
user_capabilities = response['user']['metadata']['capabilities']
```

Is it worth making an API request to auth0 every time a backend route is hit? That might be incredibly slow.

**Review by kyl191 — COMMENTED** [review]
Gating capabilities to prevent users from upgrading themselves will be added later

**Pull request merged**